File: docs/test-cases/open-redirect.py
"""Open redirect — intentionally vulnerable. DO NOT deploy."""
from flask import Flask, request, redirect

app = Flask(__name__)

# BUG: redirects to any URL the caller supplies
@app.route("/login")
def login():
    next_url = request.args.get("next", "/")
    # ... authenticate ...
    return redirect(next_url)  # attacker: /login?next=https://evil.com

# BUG: no validation on return URL
@app.route("/oauth/callback")
def oauth_callback():
    return_to = request.args.get("return_to")
    return redirect(return_to)  # open redirect via OAuth flow

# BUG: scheme-relative bypass
@app.route("/goto")
def goto():
    url = request.args.get("url")
    if not url.startswith("http://evil"):  # bypassed with //evil.com
        return redirect(url)

File: docs/test-cases/open-redirect.go
// Open redirect — intentionally vulnerable. DO NOT deploy.
package main

import "net/http"

// BUG: redirects to any URL from query parameter
func loginHandler(w http.ResponseWriter, r *http.Request) {
	next := r.URL.Query().Get("next")
	// ... authenticate ...
	http.Redirect(w, r, next, http.StatusFound) // open redirect
}

// BUG: no validation on return_to parameter
func callbackHandler(w http.ResponseWriter, r *http.Request) {
	returnTo := r.URL.Query().Get("return_to")
	http.Redirect(w, r, returnTo, http.StatusFound) // attacker: ?return_to=//evil.com
}

func main() {
	http.HandleFunc("/login", loginHandler)
	http.HandleFunc("/callback", callbackHandler)
	http.ListenAndServe(":8080", nil)
}
